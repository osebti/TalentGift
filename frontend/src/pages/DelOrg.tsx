//#region import
import config from "../../../config";
// import { useState, useEffect, Dispatch, SetStateAction } from "react";
import Modal from "@/components/Modal";
import { setGlobalState, useGlobalState } from "@/state/global";
import { Link, useNavigate } from "react-router-dom";
import Button from "@/components/Button";
//#endregion

export default function DelOrg() {
  const [user] = useGlobalState("user");
  const [org] = useGlobalState("org");
  const oid = org.oid;
  const navigate = useNavigate();

  /**
   * This function sends a request to the server
   * to delete the organization
   * @param id The ID of the organization to delete
   */
  const deleteOrgOnClick = (id: string) => {
    fetch(`http://${config.hostname}/organization/${id}`, {
      method: "DELETE",
    })
      .then((res) => res.json())
      .then((data) => {
        if (data.status === 0) {
          // User loses manager status on organization deletion
          setGlobalState("user", { ...user, manager: false });
          navigate("/success", {
            state: {
              message: "Organization deleted!",
              returnAddress: "/dashboard",
              returnName: "Dashboard",
            },
          });
        } else {
          navigate("/error", {
            state: {
              message: "Failed to delete organization ...",
              returnAddress: "/dashboard",
              returnName: "Dashboard",
            },
          });
        }
      });
  };

  return (
    <div className="w-full h-full flex justify-center items-center px-4">
      <div className="flex flex-col w-2/3 justify-center items-center gap-2">
        {/** Title and subtitle of the page */}
        <div
          id="content-header"
          className="font-semibold text-2xl flex flex-col justify-center items-center my-6 md:my-8 lg:my-10"
        >
          Are You Sure? This Action Cannot Be Undone
          <div
            id="content-title"
            className="font-semibold text-2xl md:text-3xl lg:text-4xl text-black"
          ></div>
        </div>
        <Button
          id="del-org"
          label="Delete Organization"
          onClick={() => {
            (
              document.getElementById("delete-org-modal") as HTMLDialogElement
            ).showModal();
          }}
          style="btn-secondary rounded-2xl mx-4 md:mx-8"
        ></Button>
        {/** Organization deletion modal */}
        <Modal
          id="delete-org"
          header="Delete Organization"
          message={`Are you sure you want to delete the organization with oid ${oid}?`}
          onConfirm={() => {
            deleteOrgOnClick(oid);
          }}
        />
        {/** Back to dashboard link */}
        <Link className="mx-4 md:mx-8 lg:mx-0" to="/dashboard">
          <div className="flex justify-center items-center">
            <div className="text-secondary hover:text-black transition ease-in-out text-center">
              I change my mind, back to Dashboard
            </div>
          </div>
        </Link>
      </div>
    </div>
  );
}
