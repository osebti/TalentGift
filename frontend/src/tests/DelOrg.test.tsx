//#region imports
import { cleanup, fireEvent, render, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import fetchMock from "jest-fetch-mock";

import DelOrg from "@/pages/DelOrg";
import { BrowserRouter } from "react-router-dom";
//#endregion

//#region test setup
fetchMock.enableMocks();
console.error = jest.fn();
console.warn = jest.fn();
console.log = jest.fn();

beforeEach(() => {
  fetchMock.resetMocks();
});

afterEach(cleanup);
//#endregion

//#region functions and objects used for testing
function renderDelOrgPage() {
  return (
    render(
      <BrowserRouter>
        <DelOrg/>
      </BrowserRouter>
    )
  );
}
//#endregion

//#region tests
describe("Rendering the Delete Organization Page", () => {

  it("renders correctly", () => {
    const page = renderDelOrgPage();
    expect(page).toMatchSnapshot();
  });

  it("renders the delete organization button", () => {
    const { queryByTestId } = renderDelOrgPage();
    expect(queryByTestId("del-org-button")).not.toBeNull();
  });

});

describe("Deleting the Organization", () => {
  
  it("organization is deleted if the user confirms", async () => {
    const { queryByTestId } = renderDelOrgPage();

    // mock organization deletion
    fetchMock.mockResponseOnce(
      JSON.stringify({ status: 0 }), 
      { status: 200, headers: { 'Content-Type': 'application/json' } }
    );

    // click the delete organization button
    const deleteOrgButton = queryByTestId("del-org-button");
    if (deleteOrgButton != null) {
      fireEvent.submit(deleteOrgButton);
    }

    // click confirm
    const confirmButton = queryByTestId("Confirm-button");
    if (confirmButton != null) {
      fireEvent.click(confirmButton);
    }

    // check for redirect and fetch request
    await waitFor(() => {
      expect(fetchMock).toHaveBeenCalledTimes(1);
      expect(global.window.location.href).toContain("/success");
    });
  });

  it("organization is not deleted if the user cancels", async () => {
    // mock organization deletion
    fetchMock.mockResponseOnce(
      JSON.stringify({ status: 0 }), 
      { status: 200, headers: { 'Content-Type': 'application/json' } }
    );
    // click the delete organization button
    const { queryByTestId } = renderDelOrgPage();
    const deleteOrgButton = queryByTestId("del-org-button");
    if (deleteOrgButton != null) {
      fireEvent.submit(deleteOrgButton);
    }
    // click cancel
    const cancelButton = queryByTestId("Cancel-button");
    if (cancelButton != null) {
      fireEvent.click(cancelButton);
    }
    // make sure delete organization request was not made
    await waitFor(() => {
      expect(fetchMock).toHaveBeenCalledTimes(0);
    });
  });

});
//#endregion