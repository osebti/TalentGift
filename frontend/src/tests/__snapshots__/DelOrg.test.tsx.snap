// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering the Delete Organization Page renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="w-full h-full flex justify-center items-center px-4"
      >
        <div
          class="flex flex-col w-2/3 justify-center items-center gap-2"
        >
          <div
            class="font-semibold text-2xl flex flex-col justify-center items-center my-6 md:my-8 lg:my-10"
            id="content-header"
          >
            Are You Sure? This Action Cannot Be Undone
            <div
              class="font-semibold text-2xl md:text-3xl lg:text-4xl text-black"
              id="content-title"
            />
          </div>
          <button
            class="btn btn-md capitalize font-normal text-white text-sm md:text-base btn-secondary rounded-2xl mx-4 md:mx-8"
            data-testid="del-org-button"
            id="del-org-button"
          >
            Delete Organization
          </button>
          <dialog
            class="modal"
            data-testid="delete-org-modal"
            id="delete-org-modal"
          >
            <div
              class="modal-box border-2 border-primary"
            >
              <h3
                class="font-bold text-lg"
                id="modal-header"
              >
                Delete Organization
              </h3>
              <p
                class="py-4"
                id="modal-message"
              >
                Are you sure you want to delete the organization with oid null?
              </p>
              <div
                class="modal-action"
              >
                <form
                  method="dialog"
                >
                  <div
                    class="flex gap-3"
                    id="modal-button-container"
                  >
                    <button
                      class="btn btn-md capitalize font-normal text-white text-sm md:text-base btn btn-outline btn-accent rounded-xl"
                      data-testid="Cancel-button"
                      id="Cancel-button"
                    >
                      Cancel
                    </button>
                    <button
                      class="btn btn-md capitalize font-normal text-white text-sm md:text-base btn btn-outline btn-secondary rounded-xl"
                      data-testid="Confirm-button"
                      id="Confirm-button"
                    >
                      Confirm
                    </button>
                  </div>
                </form>
              </div>
            </div>
          </dialog>
          <a
            class="mx-4 md:mx-8 lg:mx-0"
            href="/dashboard"
          >
            <div
              class="flex justify-center items-center"
            >
              <div
                class="text-secondary hover:text-black transition ease-in-out text-center"
              >
                I change my mind, back to Dashboard
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="w-full h-full flex justify-center items-center px-4"
    >
      <div
        class="flex flex-col w-2/3 justify-center items-center gap-2"
      >
        <div
          class="font-semibold text-2xl flex flex-col justify-center items-center my-6 md:my-8 lg:my-10"
          id="content-header"
        >
          Are You Sure? This Action Cannot Be Undone
          <div
            class="font-semibold text-2xl md:text-3xl lg:text-4xl text-black"
            id="content-title"
          />
        </div>
        <button
          class="btn btn-md capitalize font-normal text-white text-sm md:text-base btn-secondary rounded-2xl mx-4 md:mx-8"
          data-testid="del-org-button"
          id="del-org-button"
        >
          Delete Organization
        </button>
        <dialog
          class="modal"
          data-testid="delete-org-modal"
          id="delete-org-modal"
        >
          <div
            class="modal-box border-2 border-primary"
          >
            <h3
              class="font-bold text-lg"
              id="modal-header"
            >
              Delete Organization
            </h3>
            <p
              class="py-4"
              id="modal-message"
            >
              Are you sure you want to delete the organization with oid null?
            </p>
            <div
              class="modal-action"
            >
              <form
                method="dialog"
              >
                <div
                  class="flex gap-3"
                  id="modal-button-container"
                >
                  <button
                    class="btn btn-md capitalize font-normal text-white text-sm md:text-base btn btn-outline btn-accent rounded-xl"
                    data-testid="Cancel-button"
                    id="Cancel-button"
                  >
                    Cancel
                  </button>
                  <button
                    class="btn btn-md capitalize font-normal text-white text-sm md:text-base btn btn-outline btn-secondary rounded-xl"
                    data-testid="Confirm-button"
                    id="Confirm-button"
                  >
                    Confirm
                  </button>
                </div>
              </form>
            </div>
          </div>
        </dialog>
        <a
          class="mx-4 md:mx-8 lg:mx-0"
          href="/dashboard"
        >
          <div
            class="flex justify-center items-center"
          >
            <div
              class="text-secondary hover:text-black transition ease-in-out text-center"
            >
              I change my mind, back to Dashboard
            </div>
          </div>
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
